modules/apps/depot/depot-web/src/main/resources/META-INF/resources/js/ConnectedSiteDropdownDefaultEventHandler.es.js

```diff
@@ -12,20 +12,25 @@
  * details.
  */
 
-import {DefaultEventHandler} from 'frontend-js-web';
+export default function getProps() {
```

Comment:
Would we be able to enable the API to accept an object or a function? I can imagine this being an easily missable "gotcha" for people who try to adopt this pattern.

```js
export default {
    onClick: // ...
}

export default function getProps() {
    return {
        onClick: // ...
    }
}

Reply:
Should be an easy enough check, yeah... I imagine the second to be needed when the props are not static and need some dynamic data like namespace to properly.

Would you think this as a good approach? To take both?

Reply:
I don't think it's a good idea to provide two ways of doing something *just because you can*. It brings complexity. In general, the less polymorphism in the types that an API accepts/returns, the easier it is to reason about.

Now, if you have a use case that the additional flexibility would *enable* (ie. that wouldn't be possible otherwise), or that becomes *considerably easier* as a result, then that would be another story.

Reply:
So, the case of us passing `namespace` and other backend generated values to the `additionalProps` method is an existing use case that is trickier the other way around.

We can also do the pure object export and create a closure with injected implicit vars... but then there would be no reference to where they're coming from other than a more than likely necessary `/* eslint-globals: namespace, foo, var */` in those that need additional params.

Using the `method` approach, at least what it takes gets codified in the signature...

Reply:
@wincent I agree that we shouldn't provide two ways "just because we can". Maybe my original question should have been "why are we exporting a function?". Since we are pointing to a file path, I don't think it makes sense to use a function unless we intend to leverage it. So I would think it would make more sense to have it be an object. I can see an additional use case of being a function if we intend to pass args to it.

Personally, I think when it comes to passing config objects around, I think it's helpful to allow for both functions that return an object and an object by itself, which is why I mentioned it. I also see this as a common pattern in the js world when passing configs around.

tldr;
Makes more sense to be for it to be an Object and not a function that returns an Object.

Reply:
>  I imagine the second to be needed when the props are not static and need some dynamic data like namespace to properly work.

How would we pass args for it to be dynamic? I'd be curious to see an example of how we would do this.

Reply:
I don't think that just because you see it a lot in the JS world that makes it a good pattern. For example, as TypeScript use becomes more and more pervasive, people are probably going to start moving away from dynamic patterns towards static ones (because when you have a polymorphic type, writing the type becomes harder, reading it becomes harder, and things like IDE type tooltips become unwieldy etc). We may find that the patterns of the future don't resemble the patterns of the past. 

And there's also the question of what "level" an API is. It's a lot more appropriate for a very high level tool like Babel, for example, to be super flexible and accept config in more ways (eg. static objects, functions that return static objects, functions that return functions etc) than for a lower level API that would be easier to reason about if it were more "primitive".

The other thing to bear in mind about functions vs objects in general is that functions are lazy (only evaluated when called) whereas static objects must be evaluated eagerly in order to be exported. (Not so relevant here, but in the abstract, pure functions are testable under a range of inputs; objects obviously are not). There are advantages/disadvantages to each, and when you choose to have *both*, that means you get *all* the advantages, and *all* the disadvantages, plus the complexity of managing both.

Reply:
@wincent what is the benefit of having this configuration of props be a function instead of an object?

I think a pure object is more preferable because I am imagining that non-FE people will run into _forgetting to return the object in the function_ or _forgetting to have a function at all_. Both of these small errors can easily come from copy/pasting a config, and there won't be a clear reason why their props aren't working, both at compile and run time.

My goal in this is to reduce bits of friction for non-FE focused devs.

Reply:
Hey @bryceosterhaus, @wincent!

```javascript
getAdditionalProps({namespace}) {
    return {
        onClick: (event) => {
            const eventName = `${namespace}selectTranslations`;
            const rootNode = document.getElementById(`#${namespace}SearchContainer`);
            ...       
        }
    }
}
```

Namespacing stuff will be the most common use case. In general we could argue that `const NAMESPACE = '_com_liferay_foo';` should do the trick for pure objects... but the thing is namespaces are dynamic for instanceable widgets (when more than one of the same type can be added to a page), so you won't know the namespace ahead of time.

Another probably common use case would be URL passing. We have some client-side facilities, but they're not on par with the ones server-side, and we don't have access to the same elements (`Request`, `Response`). So, we could start moving URL generation to JS as well, but I'd expect this to be rather common:

```javascript
import {fetch} from 'frontend-js-web';
getAdditionalProps({namespace, somethingURL}) {
    return {
        onClick: (event) => {
            fetch(somethingURL).then(...);    
        }
    }
}
```

It's  not that this makes me super happy, so that's the reason I sent it in advance, to see what we come up with. I think functions will be needed at some point... but we could also have different APIs for this:
- `additionalProps` (Object)
- `calculateAdditionalProps` (Function)

🤷

Reply:
> @wincent what is the benefit of having this configuration of props be a function instead of an object?

I wasn't arguing about function *versus* object (although we could totally explore the pros on cons of each) — I was arguing against accepting function *or* object without a clear benefit to supporting both. It seems that I misunderstood your argument all along, and you *only* want object based config.

I think @jbalsas raised some problems with the object-only approach in [his comment up above](https://github.com/liferay-frontend/liferay-portal/pull/2#discussion_r440364564).

Reply:
> but we could also have different APIs for this:
> 
> * `additionalProps` (Object)
> * `calculateAdditionalProps` (Function)
> 
> 🤷

I don't like the idea of expanding the API surface area, and I don't think that the concern that @bryceosterhaus raised ie.

> imagining that non-FE people will run into *forgetting to return the object in the function* or *forgetting to have a function at all*.

is enough to justify it.

It would mean that we'd increase our surface area (two APIs to maintain, document, test, reason about) just because we *hypothesize* that non-frontend people can't learn a simple pattern. I think this underestimates the ability of non-frontend people writing JS at Liferay. And if we can't teach people who are writing JS to be comfortable with writing a function that returns an object, then we've hired the wrong people (either because they are too stupid to learn, or we're too stupid to teach).

I know it may seem like I am nit-picking, but the road to hell is paved with good intentions. I'm worried about our systems become unmaintainable not because of one tiny little API design decision like this, but because of the accumulative effective of hundreds and thousands of such decisions accruing over many years.

Reply:
> [...] but because of the accumulative effective of hundreds and thousands of such decisions accruing over many years.

Dead by a thousand cuts ;)

Reply:
> you only want object based config.

I prefer object if we don't intend to pass any args to the function. If they intend to be dynamic, I totally get that we would need a function for it.

Personally, I still prefer allowing for both a function that returns an object or a static object. I get that it expands out API surface, but to me it doesn't seem like drastically enlarged footprint. Im js, it would be something similar to adding a one-liner like this.

```js
config = typeof config === 'function' ? config(someParams) : config
```

In the end, I don't actually care a whole lot, just trying to defend my case here. So go with what you guys think is best. I still vote for both though 😃


> just because we hypothesize that non-frontend people can't learn a simple pattern. I think this underestimates the ability of non-frontend people writing JS at Liferay.

It's not that I think they don't have the ability. I bring it up because I have helped at least 2 people at liferay with debugging a config that was malformed because they forgot to return the config function. I'm sure they never made the mistake again, but it likely took them a handful of time debugging and then finally asking me about it.


> accumulative effective of hundreds and thousands of such decisions accruing over many years

Totally agree. We also run the risk of wasting countless hours arguing over trivial patterns 😂. I do think this is a good mindset to have though, which is why I always complain when we make concessions to satisfy the source formatter.

Reply:
See some examples of `propsTransformer`:
- https://github.com/liferay-frontend/liferay-portal/pull/2/files#diff-bf360f56a310b6e0c93b1a49c9f96538R58
- https://github.com/liferay-frontend/liferay-portal/pull/2/files#diff-52d1f42c5cafa7f62c77e968a45e3869R29